### PR TITLE
feat: add support for signed commits in gitfs

### LINKED ISSUE #3053: Supported signed gitFs requests
Signing of git commits by user @renovate-bot is supported via GitHub's API but not yet via git itself. We should add the capability to [gitFs's commitFilesToBranch function](https://github.com/renovatebot/renovate/blob/07f32b35391857f748a164a738e484361dcd3401/lib/platform/git/storage.js#L260-L292).

### PR DESCRIPTION
**Pull request by rishabh1403**
Closes #3053

### CODE REVIEW

**Comment by rishabh1403** [comment]
@rarkins please review.

**Comment by rarkins** [comment]
What about configuring it globally rather than locally? config.localdir will likely change per-repository soon so that we can persist git clones between runs

**Comment by rishabh1403** [comment]
> What about configuring it globally rather than locally? config.localdir will likely change per-repository soon so that we can persist git clones between runs

If we configure globally, won't it affect all the repos? A single private key for all repos? @rarkins

**Comment by rarkins** [comment]
Yes that’s right. It’s an admin option and assumed to be same for all repos.

**Comment by rishabh1403** [comment]
> Yes that’s right. It’s an admin option and assumed to be same for all repos.

So the directory would just be `/` i.e. the root directory in that case, just to be clear?

**Comment by rarkins** [comment]
Actually sorry - I changed my mind. We should not run the git config globally as it could interfere with someone's system if they run it in their own user space. It should instead be run each time as part of *repository* worker in the repository's working directory, so that it only applies its setting locally.

**Comment by rishabh1403** [comment]
@rarkins please review.

File: lib/workers/repository/index.js [hunk 1]
```diff
@@ -14,6 +15,23 @@ module.exports = {
   renovateRepository,
 };
 
+async function setGitPrivateKey(config) {
+  if (config.gitPrivateKey && config.localDir) {
```
Comment (rarkins) [inline lib/workers/repository/index.js]: Is there any scenario where config.localDir *wouldn't* be set?

File: lib/workers/repository/index.js [hunk 2]
```diff
@@ -14,6 +15,23 @@ module.exports = {
...
+    try {
+      const cwd = config.localDir;
+      const git = Git(cwd).silent(true);
+      await git.raw([
+        'config',
+        '--global',
```
Comment (rarkins) [inline lib/workers/repository/index.js]: Not global. We want this git config setting to be run locally so it doesn't affect the user's own git settings

File: lib/workers/repository/index.js [hunk 3]
```diff
@@ -31,6 +49,7 @@ async function renovateRepository(repoConfig) {
       config.localDir || path.join(tmpDir, config.platform, config.repository);
     await fs.ensureDir(config.localDir);
     logger.debug('Using localDir: ' + config.localDir);
+    await setGitPrivateKey(config);
```
Comment (rarkins) [inline lib/workers/repository/index.js]: Do you think we could do this as part of platform.initRepo instead? That way we do it immediately after we clone or fetch and be sure it's done before any commits.

File: lib/workers/repository/index.js (hunk 1 above)
Comment (rishabh1403) [reply lib/workers/repository/index.js]: It's just a safety check, if by any chance it's not present the operation won't execute

File: lib/workers/repository/index.js (hunk 2 above)
Comment (rishabh1403) [reply lib/workers/repository/index.js]: Got it :+1:

File: lib/workers/repository/index.js (hunk 3 above)
Comment (rishabh1403) [reply lib/workers/repository/index.js]: I found that `localDir` is ensured here, so I made the function call over here to make sure localDir is present

Comment (rarkins) [reply lib/workers/repository/index.js]: But don't we call platform.initRepo after this too? i.e. we can be assured its done after the localDir is set

Comment (rishabh1403) [reply lib/workers/repository/index.js]: yeah, you're right, makes sense

Comment (rishabh1403) [reply lib/workers/repository/index.js]: changed the location

File: lib/workers/repository/init/apis.js [hunk 4]
```diff
@@ -7,6 +8,18 @@ function assignPlatform(config) {
   return config;
 }
 
+async function setGitPrivateKey(config) {
```
Comment (rarkins) [inline lib/workers/repository/init/apis.js]: I'd like to move this initialization code out of workers/repository and into gitFs's initRepo function: https://github.com/renovatebot/renovate/blob/14214835a116d79007d7dc2bd3f4913eb73cdbea/lib/platform/git/storage.js#L53

**Review by rarkins — CHANGES_REQUESTED** [review]

**Pull request closed without merging**